src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1242,19 +1246,192 @@
+            :test (handle-expr prev
+                               (compile-test (:constraints condition))
+                               id
+                               :test-expr
+                               {:compile-ctx {:condition condition
+                                              :env (:env beta-node)
+                                              :msg "compiling test node"}})
+            :accumulator (cond-> (handle-expr prev
+                                              (compile-accum (:accumulator beta-node) (:env beta-node))
+                                              id
+                                              :accum-expr
+                                              {:compile-ctx {:condition condition
+                                                             :accumulator (:accumulator beta-node)
+                                                             :env (:env beta-node)
+                                                             :msg "compiling accumulator"}})
+
+                                 (:join-filter-expressions beta-node)
+                                 (handle-expr (compile-join-filter (:join-filter-expressions beta-node)
+                                                                   (:join-filter-join-bindings beta-node)
+                                                                   (:new-bindings beta-node)
+                                                                   (:env beta-node))
+                                              id
+                                              :join-filter-expr
+                                              {:compile-ctx {:condition condition
+                                                             :join-filter-expressions (:join-filter-expressions beta-node)
+                                                             :env (:env beta-node)
+                                                             :msg "compiling accumulate with join filter node"}}))
+            :query prev)))
+      exprs
+      (:id-to-condition-node beta-graph))))
+
+(sc/defn compile-exprs :- schema/NodeFnLookup
+  "Takes a map in the form produced by extract-exprs and evaluates the values(expressions) of the map in a batched manner.
+   This allows the eval calls to be more effecient, rather than evaluating each expression on its own.
+   See #381 for more details."
+  [key->expr :- schema/NodeExprLookup
+   partition-size :- sc/Int]
+  (let [batching-try-eval (fn [node-keys exprs]
+                            ;; Try to evaluate all of the expressions as a batch. If they fail the batch eval then we
+                            ;; try them one by one with their compilation context, this is slow but we were going to fail
+                            ;; anyway.
+                            (try
+                              (eval exprs)
+                              (catch Exception e
+                                (mapv (fn [expr node-keys]
```

Comment:
"doseq" seems more in line with your intent here since this return value isn't used.

Reply:
Used `mapv` to keep from having to zip the 2 lists back together, if you feel strongly i can use `doseq` instead

Reply:
Oh, I see, the anonymous function has two args.  Never mind, this is OK and it is a minor deal anyway.  It might be worth adding a quick comment that we want mapv rather than map to avoid laziness.

Do we have an existing test that validates this "rethrow if an individual expression actually is not compilable" workflow?  If not we should probably add one.